File: src/main/java/com/baulsupp/oksocial/security/CertificatePin.java
package com.baulsupp.oksocial.security;

import com.baulsupp.oksocial.output.util.UsageException;
import java.util.List;
import okhttp3.CertificatePinner;

import static java.util.stream.Collectors.groupingBy;

public class CertificatePin {
  private final String pattern;
  private final String pin;

  public CertificatePin(String patternAndPin) {
    String[] parts = patternAndPin.split(":", 2);

    pattern = parts[0];
    pin = parts.length == 2 ? parts[1] : null;
  }

  public String getPattern() {
    return pattern;
  }

  public String getPin() {
    if (pin == null) {
      throw new UsageException(
          "--certificatePin expects host:AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA");
    }

    return pin;
  }

  public static CertificatePinner buildFromCommandLine(List<CertificatePin> pins) {
    CertificatePinner.Builder builder = new CertificatePinner.Builder();

    pins.stream().collect(groupingBy(CertificatePin::getPattern)).forEach((host, pinList) -> {
      String[] pinArray = pinList.stream().map(CertificatePin::getPin).toArray(String[]::new);

      builder.add(host, pinArray);
    });

    return builder.build();
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/security/CertificatePin.kt
package com.baulsupp.oksocial.security

import com.baulsupp.oksocial.output.util.UsageException
import okhttp3.CertificatePinner

class CertificatePin(patternAndPin: String) {
  val pattern: String
  private val pin: String?

  init {
    val parts = patternAndPin.split(":".toRegex(), 2).toTypedArray()

    pattern = parts[0]
    pin = if (parts.size == 2) parts[1] else null
  }

  fun getPin(): String {
    if (pin == null) {
      throw UsageException(
        "--certificatePin expects host:AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")
    }

    return pin
  }

  companion object {

    fun buildFromCommandLine(pins: List<CertificatePin>): CertificatePinner {
      val builder = CertificatePinner.Builder()

      pins.groupBy { it.pattern }.forEach { (pattern, pins) ->
        builder.add(pattern, *(pins.map { it.pin }.toTypedArray()))
      }

      return builder.build()
    }
  }
}
